### ISSUE TITLE
[Feature Request]: Implement some additional yt-dlp options/commands

### ISSUE THREAD

**Issue by M4NH3X** [tubearchivist/tubearchivist#316]
### Already implemented?

- [X] I have read through the [wiki](https://github.com/tubearchivist/tubearchivist/wiki).
- [X] I understand the [scope](https://github.com/tubearchivist/tubearchivist/wiki/FAQ) of this project and am aware of the [known limitations](https://github.com/tubearchivist/tubearchivist#known-limitations).

### Your Feature Request

Currently we can't pass custom yt-dlp commands, it would be nice to have that option.

I know if we allow all options then it will break TA but yt-dlp is really robust, so that we can allow few whitelisted commands so that things won't break.

Thank you for this great project!

------

If this request is rejected for some reason, would love to see these options integrated into TA web-ui themselves. 

* [Sorting Formats:](https://github.com/yt-dlp/yt-dlp#sorting-formats) `-S` (this is far more robust than -f)
* Extract audio: `-x` `--audio-format` (Channels dedicated to podcasts, etc. can be really helpful)
* `-N` or `--concurrent-fragments` (some formats use fragments)
* `--write-info-json` `--write-description` (it will increase portability of downloaded data if users plans to migrate)

That's all options I can think of now which won't break things but increase flexibility.

### Your help is needed!

- [ ] Yes I can help with this feature request!

**Comment by bbilly1** [comment]
So with arbitrary settings import it's going to be very easy to break Tube Archivist. As that is in essence untestable. So it's not a question if yt-dlp is robust, which it is, it's that Tube Archivist expects certain files to be there, and other not to be there. Also depending on the option you choose, you'll want to index things differently, but that wouldn't be possible with arbitrary settings. 

I guess that's the difference between a download GUI that only needs to worry about that once at download, and a media server where everything needs to fit with each other indefinitely. 

But for your configuration options:
1. Sorting formats: Sounds good, that's only at download time, so that's straight forward to implement.
2. Extract audio: That's on the roadmap with the podcast rss feed, so that can be implemented together with that.
3. `-N`: I'm not familiar with that one, but sounds like a download only option? Need to look into that.
4. Write description is implemented with `--embed-metadata`, which also embeds description plus more. I think there is also `--embed-info-json`? Does that work with mp4 files? I'd rather not create separate files for that, as this will require more changes to handle that for things like *rescan filesystem* and the like and will double the amount of files you create, that's why you have a media server! Plus embedded things are easy enough to extract if portability is your concern. 

How does that sound?

**Comment by DonneyF** [comment]
I have "implemented" this feature by overriding certain source files in the provided docker containers.

i.e. adding `-v /opt/tubearchivist/yt_dlp_base.py:/app/home/src/download/yt_dlp_base.py` to `docker run` or similar for other uses of docker.

## My particular use-case

In yt-dlp `2022.10.04`, support was added to change the extraction language for YouTube. This allows users to download videos where the metadata that is downloaded is translated to the desired language, wherever the uploader has made such a translation.

Example links:
https://www.youtube.com/watch?v=XZnySZWeSJM
https://www.youtube.com/watch?v=dmfNXOrqU4w
https://www.youtube.com/watch?v=-_2nWMF20y0
https://www.youtube.com/watch?v=2sKgVz79NZE
https://www.youtube.com/watch?v=04C3ll-BaEw

These videos have been translated to English by the uploaders themselves. To see this in action, you pass in `--extractor-args "youtube:lang=en"` to the yt-dlp CLI, which is what I want to pass to TubeArchvist.

Do do this using the yt-dlp Python API, I added the following to the `YT_OBS` dictionary of the `YtWrap` object in `/home/src/download/yt_dlp_base.py`:

```
        'extractor_args': {
            'youtube': {
                'lang': ['en']
            }
        }
```

I saved the file arbitrarily to `/opt/tubearchivist/yt_dlp_base.py` and then use docker volumes as above. A restart of the container is necessary for the changes to take effect.

This particular implementation forces these arguments to be "global". If you wanted to add arguments specifically to downloading or just for the queue you could add additional functionality to do so, or override the other OBS dictionaries in the other yt-dlp API interfacing methods.

### Results
<details>
<summary>Download Queue</summary>

![screencapture-10-0-0-100-8000-downloads-2022-10-26-15_09_40](https://user-images.githubusercontent.com/1490651/198495040-acc1b51f-00b6-4ba8-b51e-6793ef907c6e.jpg)
![screencapture-10-0-0-100-8000-downloads-2022-10-26-15_14_07](https://user-images.githubusercontent.com/1490651/198495044-b0296fa8-edbc-4135-97c2-0f277c968aff.jpg)

</details>
<details>
<summary>Video Player</summary>

![screencapture-10-0-0-100-8000-video-04C3ll-BaEw-2022-10-26-15_12_26](https://user-images.githubusercontent.com/1490651/198495046-3944201b-b603-4da0-9455-ffbd68f90ac8.jpg)
![screencapture-10-0-0-100-8000-video-04C3ll-BaEw-2022-10-26-15_15_04](https://user-images.githubusercontent.com/1490651/198495049-138b22ee-9fbb-4655-bda0-63f2da2bedc6.jpg)

</details>

Seems to work fine. Search is fine and nothing out of the ordinary in the logs.

**Comment by pukkandan** [comment]
I think custom options should be allowed, possibly with a warning that it's an advanced feature and the user is "on their own". yt-dlp has so many options, that it would be impossible to add them all to the GUI. Some examples that are infeasible to do via GUI include `--parse-metadata`, `--extractor-args`, `--ppa` etc.

You can provide a field where users enter CLI switches and use `yt_dlp.parse_options(shlex.split(opts)).ydl_opts` to convert them into their API equivalent

**Comment by bbilly1** [comment]
@dsander thanks a lot for getting the ball rolling here! Please keep going! :-)

I think the `-N` option translates to `concurrent_fragment_downloads`, I was thinking to hardcode this, probably not a lot of benefit to expose this to users, but if we have it set to maybe 4 (?) this should be a good for most people and will speed up download for these cases while not harm low end hardware too much...

Then the extractor language as outlined [above](https://github.com/tubearchivist/tubearchivist/issues/316#issuecomment-1294403586) might work similar as the format-sort option you have already implemented, if you want to give this a try?

**Comment by bbilly1** [comment]
that is now merged in v0.3.6, please update. This includes --format-sort and extractor language. 

Unfortunately `-N` or *concurrent fragment download* was breaking our task queue handling and notifications. yt-dlp will create multiple threads in our already multi threaded task queue, which keeps breaking things. I decided to not implement this for now, also because it was reporting inaccurate progress and speeds as only one thread will send notifications.
